engines/made/made.cpp

```diff
@@ -189,62 +191,70 @@ void MadeEngine::handleEvents() {
 		case Common::EVENT_RBUTTONUP:
 			_eventNum = 3;
 			break;
-
 		case Common::EVENT_KEYDOWN:
-			// Handle any special keys here
-			// Supported keys taken from https://web.archive.org/web/20141114142447/http://www.allgame.com/game.php?id=13542&tab=controls
-
-			switch (event.kbd.keycode) {
-			case Common::KEYCODE_KP_PLUS:	// action (same as left mouse click)
-				_eventNum = 1;		// left mouse button up
-				break;
-			case Common::KEYCODE_KP_MINUS:	// inventory (same as right mouse click)
-				_eventNum = 3;		// right mouse button up
-				break;
-			case Common::KEYCODE_UP:
-			case Common::KEYCODE_KP8:
-				_eventMouseY = MAX<int16>(0, _eventMouseY - 1);
-				g_system->warpMouse(_eventMouseX, _eventMouseY);
-				break;
-			case Common::KEYCODE_DOWN:
-			case Common::KEYCODE_KP2:
-				_eventMouseY = MIN<int16>(199, _eventMouseY + 1);
-				g_system->warpMouse(_eventMouseX, _eventMouseY);
-				break;
-			case Common::KEYCODE_LEFT:
-			case Common::KEYCODE_KP4:
-				_eventMouseX = MAX<int16>(0, _eventMouseX - 1);
-				g_system->warpMouse(_eventMouseX, _eventMouseY);
+		//	// Handle any special keys here
+		//	// Supported keys taken from https://web.archive.org/web/20141114142447/http://www.allgame.com/game.php?id=13542&tab=controls
+			if (event.kbd.keycode == Common::KEYCODE_BACKSPACE) {
+				_eventNum = 5;
+				_eventKey = 9;
+			} else {
+				_eventNum = 5;
+				_eventKey = event.kbd.ascii;
+			}
+			break;
+		case Common::EVENT_CUSTOM_ENGINE_ACTION_START:
+			_action = (MADEAction)event.customType;
+			switch (event.customType) {
+			case kActionMenu:
+				_eventNum = 5;
+				_eventKey = 21;
 				break;
-			case Common::KEYCODE_RIGHT:
-			case Common::KEYCODE_KP6:
-				_eventMouseX = MIN<int16>(319, _eventMouseX + 1);
-				g_system->warpMouse(_eventMouseX, _eventMouseY);
+			case kActionSaveGame:
+				_eventNum = 5;
+				_eventKey = 22;
 				break;
-			case Common::KEYCODE_F1:		// menu
-			case Common::KEYCODE_F2:		// save game
-			case Common::KEYCODE_F3:		// load game
-			case Common::KEYCODE_F4:		// repeat last message
+			case kActionLoadGame:
 				_eventNum = 5;
-				_eventKey = (event.kbd.keycode - Common::KEYCODE_F1) + 21;
+				_eventKey = 23;
 				break;
-			case Common::KEYCODE_BACKSPACE:
+			case kActionRepeatMessage:
 				_eventNum = 5;
-				_eventKey = 9;
+				_eventKey = 24;
 				break;
 			default:
-				_eventNum = 5;
-				_eventKey = event.kbd.ascii;
 				break;
 			}
 			break;
-
+		case Common::EVENT_CUSTOM_ENGINE_ACTION_END:
+			_action = kActionNone;
+			break;
 		default:
 			break;
 
 		}
 	}
 
+	switch (_action) {
```

Comment:
Could this be handled with the other actions EVENT_CUSTOM_ENGINE_ACTION_START?

Reply:
So I introduced _action because EVENT_CUSTOM_ENGINE_ACTION_START does not trigger each loop when you press down the key like how keydown does, it triggers once when you press the key and EVENT_CUSTOM_ENGINE_ACTION_END is triggered once when the key is left

Reply:
Ah, so we're intentionally repeating until key release or overridden by another action. Perhaps a comment indicating that would be useful.

Reply:
Commented

Reply:
Does enabling `allowKbdRepeats();` for the cursor actions not work as expected?

Reply:
Haven't tried that method, looking into it

Reply:
It workds with allowKbdRepeats();